### ISSUE TITLE
BUG: matplotlib 3.4.0 -- Scatter with colormap and legend gives TypeError: object of type 'NoneType' has no len()

### ISSUE THREAD

**Issue by pllim** [matplotlib/matplotlib#19779]
### Bug report

**Bug summary**

With `matplotlib` 3.4.0, scatter plot with colormap and legend gives `TypeError: object of type 'NoneType' has no len()`.

**Code for reproduction**

```python
import matplotlib.pyplot as plt
y = [1, 2, 3, 4, 5]
plt.scatter(y, y, label='Y', lw=0, s=8, c=y, cmap='viridis')
plt.legend()
plt.show()
```

**Actual outcome**

```
Traceback (most recent call last):
  File ".../matplotlib/backends/backend_qt5.py", line 475, in _draw_idle
    self.draw()
  File ".../matplotlib/backends/backend_agg.py", line 406, in draw
    self.figure.draw(self.renderer)
  File ".../matplotlib/artist.py", line 74, in draw_wrapper
    result = draw(artist, renderer, *args, **kwargs)
  File ".../matplotlib/artist.py", line 51, in draw_wrapper
    return draw(artist, renderer, *args, **kwargs)
  File ".../matplotlib/figure.py", line 2734, in draw
    mimage._draw_list_compositing_images(
  File ".../matplotlib/image.py", line 132, in _draw_list_compositing_images
    a.draw(renderer)
  File ".../matplotlib/artist.py", line 51, in draw_wrapper
    return draw(artist, renderer, *args, **kwargs)
  File ".../matplotlib/_api/deprecation.py", line 431, in wrapper
    return func(*inner_args, **inner_kwargs)
  File ".../matplotlib/axes/_base.py", line 2925, in draw
    mimage._draw_list_compositing_images(renderer, self, artists)
  File ".../matplotlib/image.py", line 132, in _draw_list_compositing_images
    a.draw(renderer)
  File ".../matplotlib/artist.py", line 51, in draw_wrapper
    return draw(artist, renderer, *args, **kwargs)
  File ".../matplotlib/legend.py", line 614, in draw
    self._legend_box.draw(renderer)
  File ".../matplotlib/offsetbox.py", line 368, in draw
    c.draw(renderer)
  File ".../matplotlib/offsetbox.py", line 368, in draw
    c.draw(renderer)
  File ".../matplotlib/offsetbox.py", line 368, in draw
    c.draw(renderer)
  [Previous line repeated 1 more time]
  File ".../matplotlib/offsetbox.py", line 694, in draw
    c.draw(renderer)
  File ".../matplotlib/artist.py", line 51, in draw_wrapper
    return draw(artist, renderer, *args, **kwargs)
  File ".../matplotlib/collections.py", line 1009, in draw
    super().draw(renderer)
  File ".../matplotlib/artist.py", line 51, in draw_wrapper
    return draw(artist, renderer, *args, **kwargs)
  File ".../matplotlib/collections.py", line 380, in draw
    len(facecolors) == 1 and len(edgecolors) == 1 and
TypeError: object of type 'NoneType' has no len()
```

**Expected outcome**

With matplotlib 3.3.4, I get this (no warning or error):

![Screenshot 2021-03-26 141815](https://user-images.githubusercontent.com/2090236/112675923-28e2c980-8e3e-11eb-9954-9eda46cbb6ca.jpg)

**Matplotlib version**

  * Operating system: Red Hat Enterprise Linux Server release 7.9 (Maipo)
  * Matplotlib version (`import matplotlib; print(matplotlib.__version__)`): 3.4.0
  * Matplotlib backend (`print(matplotlib.get_backend())`): Qt5Agg
  * Python version: 3.8.3
  * Jupyter version (if applicable): N/A
  * Other libraries: Not sure what else you need, just let me know.

**Comment by smartlixx** [comment]
A bisect diagnose shows that the problem was introduced in the commit 
https://github.com/matplotlib/matplotlib/commit/4c3d6449249a9e4790865f2a7e825be591581958, which changed the way Collections deal with color. 

A quick fix is easy, but I don't have much in-depth understanding of this part of the code. If anyone is familiar with this part can have a discussion, that would be great. Or I just open a PR for discussion.

**Comment by dstansby** [comment]
Opening a PR is probably the best way forward 👍

**Comment by smartlixx** [comment]
> A bisect diagnose shows that the problem was introduced in the commit
> [4c3d644](https://github.com/matplotlib/matplotlib/commit/4c3d6449249a9e4790865f2a7e825be591581958) (#18480), which changed the way Collections deal with color.
> 
> A quick fix is easy, but I don't have much in-depth understanding of this part of the code. If anyone is familiar with this part can have a discussion, that would be great. Or I just open a PR for discussion.

Actually, this problem has been solved in the commit https://github.com/matplotlib/matplotlib/commit/ed6d92b00426429d3f4350fc3710155fa46160c0, but this commit was not included in v3.4.x branch. Backport that commit will solve this problem.

**Comment by QuLogic** [comment]
That commit removes API and cannot be included in v3.4.x directly.

**Comment by QuLogic** [comment]
Maybe we should just drop the `self._check_update('array')` though, @efiring?

**Comment by QuLogic** [comment]
Closed by #19816.

**Comment by pllim** [comment]
Thanks for the quick fix!
